### DISCUSSION TITLE
Glossary of Buildpacks (Align on useful buildpacks terminology)

### DISCUSSION THREAD

**Discussion by fg-j** [paketo-buildpacks/feedback#8] [Ideas]
In the Working Group meeting on 3/16/21, we discussed that there are many buildpacks concepts that don't have intuitive nouns to name them that we all agree upon. Some examples include:

1. Layers that a buildpack adds to a launch image to include an `exec.d` script that the lifecycle will run at container launch time
2. A more descriptive name for a "metabuildpack"
3. Buildpacks whose sole job is to install a dependency and put it on the `$PATH`

In this discussion, let's collect more of these "concepts that need nouns" and propose words for them. Some of these might end up in Paketo docs, some might end up as changes to the upstream Cloud Native Buildpacks specification.

**If you're proposing a concept + name, it'd be helpful if you link to concrete examples of said concept. These could be lines of code, whole buildpacks, frequently-used patterns, etc.**

**Comment by ForestEckhardt** [comment]
Another concept to consider is a buildpack that runs some build process does not produce consumable layers, something like `go-mod-vendor`, which vendors your app but only produces a cache layer which it uses on rebuilds.

**Comment by arjun024** [comment]
"Launch process buildpack" Aka "Start command buildpack" e.g. https://github.com/paketo-buildpacks/unicorn

**Reply by fg-j** [reply]
I think I like "launch process" because it's reflected in the [CNB spec](https://github.com/buildpacks/spec/blob/main/buildpack.md#launch) terminology for what happens at launch time: "...buildpack-provided process type is chosen as the start command"

Meaning that *buildpacks* provide launch processes and the *lifecycle* sets a start command

**Comment by fg-j** [comment]
We've got a bunch of buildpacks whose only job is to install a distribution or tool (as mentioned in @sophiewigmore 's [blog post](https://blog.paketo.io/posts/building-a-cloud-native-buildpack-isnt-as-hard-as-you-think/))

These include:
- [go-dist](https://github.com/paketo-buildpacks/go-dist)
- [bundler](https://github.com/paketo-buildpacks/bundler)
- [yarn](https://github.com/paketo-buildpacks/yarn)

I propose we call these *"Dependency Installation Buildpacks"*

This (hopefully) distinguishes them from buildpacks that install or vendor your app's dependencies, like 
- [bundle install](https://github.com/paketo-buildpacks/bundle-install) or 
- [go mod vendor](https://github.com/paketo-buildpacks/go-mod-vendor)

which we might call *"Vendoring Buildpacks"*

**Reply by ryanmoran** [reply]
I think _"Vendoring Buildpacks"_ may represent a subset of a larger group you might call _"Build Process Buildpacks"_. This larger group also includes buildpacks that compile your code like `go-build` or `rails-assets`.

**Comment by arjun024** [comment]
(This has a noun but useful to be documented at a central place)
* `Project Path` - as used in https://github.com/paketo-buildpacks/dotnet-publish/blob/v0.1.4/README.md#buildpackyml-configurations

**Comment by dmikusa** [comment]
I think it would help to have a term for buildpacks that do not care what stack they run on. Examples would be buildpacks that just run a tool that is on the PATH (php-composer) or most buildpacks that just end up setting a start command (i.e. launch process buildpacks -> https://github.com/paketo-buildpacks/feedback/discussions/8#discussioncomment-490392). 

Possible names: stack agnostic buildpack or noarch buildpack

**Comment by dmikusa** [comment]
I don't think we have a term to call environment variables set specifically to configure a buildpack, like BP_TOMCAT_CONTEXT_PATH or  BP_JVM_VERSION or BPL_JVM_HEAD_ROOM.

Possible names: buildpack configuration variables or buildpack setting variables

**Comment by fg-j** [comment]
On the [Paketo site](https://paketo.io/docs/buildpacks/language-family-buildpacks/#composite-buildpacks), we refer to the metabuildpacks as 
- metabuildpacks
- language family buildpacks
- composite buildpacks

**Reply by fg-j** [reply]
I think "composite buildpacks" is more clear than "metabuildpacks" .

And, I suppose, the Paketo language family buildpacks are a _type_ of composite buildpack.
